### PR TITLE
spirv-fuzz: Tolerate absent ids in data synonym fact management

### LINKED ISSUE #3952: spirv-fuzz: TransformationEquationInstruction can fail if instructions have been removed from the module
TransformationEquationInstruction can cause "corollary" facts to be added to the module based on existing facts.  If some of the base object ids associated with existing facts are gone from the module then construction of corollary facts can fail.  We either need to be more conservative about whether to add corollary facts, or be more conservative about when to assert that data descriptors are well-formed, omitting such checks in cases where the object associated with one or other data descriptor is no longer present.

### PR DESCRIPTION
**Pull request by afd**
Fixes #3952.

### CODE REVIEW

File: source/fuzz/fact_manager/data_synonym_and_id_equation_facts.cpp [hunk 1]
```diff
@@ -581,26 +587,26 @@ void DataSynonymAndIdEquationFacts::ComputeClosureOfFacts(
...
             // The objects are not both available in the module, so we cannot
             // investigate the types of the associated data descriptors; we need
             // to move on.
             continue;
           }
           // Get the type of obj_1
-          auto dd1_root_type_id = dd1_object->type_id();
+          auto dd1_root_type_id =
+              fuzzerutil::GetTypeId(ir_context_, dd1->object());
+          ;
```
Comment (paulthomson) [inline source/fuzz/fact_manager/data_synonym_and_id_equation_facts.cpp:599]: Extra semi?

File: source/fuzz/fact_manager/data_synonym_and_id_equation_facts.cpp [hunk 2]
```diff
@@ -581,26 +587,26 @@ void DataSynonymAndIdEquationFacts::ComputeClosureOfFacts(
...
           // obj_1[a_1, ..., a_m].
           auto dd1_prefix_type = fuzzerutil::WalkCompositeTypeIndices(
               ir_context_, dd1_root_type_id, dd1_prefix.index());
 
           // Similarly, get the type of obj_2 and use it to get the type of
           // obj_2[b_1, ..., b_n].
-          auto dd2_root_type_id = dd2_object->type_id();
+          auto dd2_root_type_id =
+              fuzzerutil::GetTypeId(ir_context_, dd2->object());
+          ;
```
Comment (paulthomson) [inline source/fuzz/fact_manager/data_synonym_and_id_equation_facts.cpp:609]: Extra semi?

**Review by paulthomson — APPROVED** [review]

File: source/fuzz/fact_manager/data_synonym_and_id_equation_facts.cpp (hunk 2 above)
Comment (afd) [reply source/fuzz/fact_manager/data_synonym_and_id_equation_facts.cpp]: Fixed, thanks.

File: source/fuzz/fact_manager/data_synonym_and_id_equation_facts.cpp (hunk 1 above)
Comment (afd) [reply source/fuzz/fact_manager/data_synonym_and_id_equation_facts.cpp]: Fixed, thanks.

**Pull request merged**